### PR TITLE
call API gateway instead of ZUUL gateway

### PR DESCRIPTION
**Pull request by mahsa-bytt**
this is the link of my ticket: https://app.clubhouse.io/squad2/story/17084/star-when-javascript-sdk-makes-an-api-call-it-needs-to-call-the-api-gateway-rather-than-the-zuul-gateway

in this ticket we want to call the AWS Api gateway instead of our ZUUL gateway. 
so the version1(v1) is our old way that uses mappId+bpiKey to create and send request, and the version2 is the new way that uses accessToken in the authorisation header for authorising. 
I used the overloading for constructor in OverledgerSDK.ts and Provider.ts because they should support both way of authorising.
the "balance-query-v2.js" is how we can use balance in new version for example.
the "refresh-access-token.js" is for refreshing and exchanging the old accesstoken with new one.

### CODE REVIEW

**Review by theoQuant — APPROVED** [review]
Looks good

File: examples/oAuth2.0/refresh-access-token.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
...uyzcTh40jfLI2qJNPgoi8RUH0uIqzPE2QVLbOUy5omBPVp_8wjvTls-7Um-CqUjGHeveyyDa01vItDK4AOkjxSEo7iXqKjxNUwl-SoJhb9nvwSgP-CR3KY0p8o5tfI2z6_EEeBPJxaZtjr67yTAjZBNKasESi7bY9-O4pcH6tFLnqecX7T9WFQk6bVHqxScEg0bBc0RBmC13zWS3rXI4MYVAWjdV1O1dwZyMXouNI_RVS07mjMhXLNvThHAvci3Fdh13NxQyHMBqjurHv5uXLH51vAToyT7DUvul9we1BjGvqvAmOPiMBTVYJRAHu31COQHu1y5nhdVs_NYgAmO-vwWO9T-chghXS0pXY1vn5G0hbiHyinTFWVgbU4eMyIv5ir64oAWIbGbVgFA1eBIl14_NDR4kh1IBshTgiFnIW61lslxN4a9P8GAYglFmCrSJfsAiUxLGLSjGPH5Of54z1lA9J7DcWe6Yl7mw4gTWTJst3GsAluaev2glVVFKLlLGNmjFg-ElmB8DzanGaOA74DIUoqsdAMkmMIilG1QAh5Cs__bvgdwcgKyPNPtdoSI4oVMO2nYpKxFOAWKHys5TQ737V5U6kJsF9N4vixU7PyRZT1a4mIBwYVwQZI4SvA6zvKYVzAvJrFK7SSDmTnBjKV5vCE_u3srVxNL8J4zRk72hTqpGT8yREYF5pnw2eZDZWUKPkmC2-OkgPW6rWg5M48GCr6ino8Wvg-62Xj5L9ezCkkY-5rxz1gMdOM-lNggwypSmyeD-LlzjVFK6JXKvoN5L14TF1-LsmOECjjqmN3BuDR4AFc8xOseVFTa8KToIJLhDuTca4-P1gClPH-PiojfvGzoOa1EzKruai6Qdi5WO7F78agKPFCHdL48gV8W4cBiN2PbxFLYe7onPIOivigZOmD_RHzmmVtnlzqr48Sg0132oHhMnJJOqxAXotRsqs8MYd-iALd-ki2deuolg65Mmn4QMNQgi1qIddtGRUWY75nhet5m1UNglGKpADSfmZkvUbNMYle7x0FZ96Kus2sYAVT8Axh9AisXhs2cOySF41zITlnk9j3Ka6uatpN9lDb9cYU27Km-aXOIiyS7gYCdXxT6wGmEe_38Ocug0Hy5FJza-mgWfqviy9zYgyiR6AS4CqI_zEuosxgexMtJ8OA.OwLTkwHs1pZgMt7N9v4LEw');
```
Comment (najla-quant) [inline examples/oAuth2.0/refresh-access-token.js:10]: Could you add a comment to say where this data is coming from ?

File: examples/oAuth2.0/balance-query-v2.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    ];
+
+    const balances = await overledger.getBalances(array);
+    console.log('Balances:\n', balances.data);
+    console.log("");
+
+  } catch (e) {
+    console.error('error', e.stack);
+  }
+})();
```
Comment (alexchiriac) [inline examples/oAuth2.0/balance-query-v2.js:40]: Missing newline :)

File: packages/overledger-core/src/OverledgerSDK.ts [hunk 3]
```diff
@@ -18,31 +18,59 @@ class OverledgerSDK {
...
+  /**
+* Create the Overledger SDK
+* */
+  constructor(...args: any[]) {
+    if (args.length === 3) {
+      console.log("using V1...");
+      this.mappId = args[0];
+      this.bpiKey = args[1];
+      var options = args[2];
+      this.network = options.provider && options.provider.network || TESTNET;
```
Comment (alexchiriac) [inline packages/overledger-core/src/OverledgerSDK.ts:46]: Here, I'm not sure args.length would be the correct way to differentiate between v1 and v2 calls.

Here's some background info: we use options.provider.network in the instance that a client will want to supply a custom URL for their Overledger instance. This way, our library isn't limited to interact just with our Testnet and Mainnet, but could also be used to interact with an "on-premise" installation or our Devnet.

Comment (alexchiriac) [reply packages/overledger-core/src/OverledgerSDK.ts:46]: So we need to allow the usage of v2 with three args as well, to allow the specification of a custom Overledger URL.

Comment (alexchiriac) [inline packages/overledger-core/src/OverledgerSDK.ts:46]: Couldn't there be an occasion where they provide mappId and bpiKey, but not extra options? So the args.length would be equal to 2 and jump to the else block?

**Review by alexchiriac — COMMENTED** [review]
Left a comment :)

File: packages/overledger-core/src/OverledgerSDK.ts [hunk 4]
```diff
@@ -18,31 +18,59 @@ class OverledgerSDK {
...
+        const dlt = this.loadDlt(dltConfig);
+        this.dlts[dlt.name] = dlt;
+      });
 
-    this.validateOptions(options);
+      this.provider = new Provider(this.mappId, this.bpiKey, options.provider);
+      this.request = this.provider.createRequest();
+      this.search = new OverledgerSearch(this);
+    } else if (args.length === 2) {
+      console.log("using V2...");
```
Comment (alexchiriac) [inline packages/overledger-core/src/OverledgerSDK.ts:59]: For logging, can we use log4js, like we do in the XRP connector, instead of using console.log?

**Review by najla-quant — APPROVED** [review]

File: packages/overledger-core/src/OverledgerSDK.ts [hunk 5]
```diff
@@ -18,31 +18,59 @@ class OverledgerSDK {
...
+        this.dlts[dlt.name] = dlt;
+      });
 
-    this.provider = new Provider(mappId, bpiKey, options.provider);
-    this.request = this.provider.createRequest();
-    this.search = new OverledgerSearch(this);
+      this.provider = new Provider(options.provider);
+      this.request = this.provider.createOAuthRequest(accessToken);
+    } else {
+      console.log("creating default constructor.");
```
Comment (alexchiriac) [inline packages/overledger-core/src/OverledgerSDK.ts:72]: Capitalize "Creating" please :)

File: packages/overledger-core/src/OverledgerSDK.ts (hunk 4 above)
Comment (alexchiriac) [reply packages/overledger-core/src/OverledgerSDK.ts:59]: Also capitalize "Using" :)

**Review by alexchiriac — COMMENTED** [review]
Left a couple more comments :D

File: examples/oAuth2.0/balance-query-v2.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+; (async () => {
+  try {
+
+    const overledger = new OverledgerSDK('eyJraWQiOiJ0UUtsM01LUFJuOEZJRmhCdng0Y3JETnU4MTNscVhkbVFjR05ZWmxFaWRrPSIsImFsZyI6IlJTMjU2In0.eyJzdWIiOiIyZmVjODFkYS0xYzZlLTQ1YjEtODM3OC04OGYxNDU5MGU4ZGYiLCJldmVudF9pZCI6ImVlNzNiMDJjLTA3NTAtNDRkMi04OTZiLTUwNTc5Y2NkZTc3OSIsInRva2VuX3VzZSI6ImFjY2VzcyIsInNjb3BlIjoicGhvbmUgb3BlbmlkIHJzMlwvd3JpdGUudHJhbnNhY3Rpb24gcHJvZmlsZSByczJcL3JlYWQudHJhbnNhY3Rpb24gcnMyXC9zZWFyY2gudHJhbnNhY3Rpb24gZW1haWwgcnMzXC93ZWJob29rLnJlYWQgcnMzXC93ZWJob29rLndyaXRlIiwiYXV0aF90aW1lIjoxNjEzNTcwNjcxLCJpc3MiOiJodHRwczpcL1wvY29nbml0by1pZHAudXMtZWFzdC0yLmFtYXpvbmF3cy5jb21cL3VzLWVhc3QtMl9Fc0QyT1VUREQiLCJleHAiOjE2MTM1NzQyNzEsImlhdCI6MTYxMzU3MDY3MSwidmVyc2lvbiI6MiwianRpIjoiMmYwZjdmODEtMjkwYy00ODhmLTkwNDItMjM5MGUxZDUzZjhlIiwiY2xpZW50X2lkIjoiMmQ1YzliaDEwNGZ1OXZhMnViamxrNHN1OTciLCJ1c2VybmFtZSI6Im1haHNhNSJ9.JO_YRIogQEg9Y0GzpVzNBOJWwaRDBxjbOslX44G-kEAItjiHLmAfVyOjTeoJbnBIl3oGP7IezEP9vp9gckdYssSHsAg9USAkYVzgsxeLiVIRW6irnMPQD76NfHBISSASgK87b8Kdb6NaorewgJWgu_U-thugKM0YCHbuMHBFrUDzqe1VL3oQYFiYgPcOyBPVxDqPwC5RuV1gwVoFG5M_EcWOQW5u3NIHl1zEt8bWf6ppxJ1MVRpynMJ8Pim2LebpbiK9YavBebwPVrA3-Vn6y5f4T7nK3imtTkSuFi_DVdacfNPIFSvCNLoB-uiewyq8YMsarPPOmWxJ384ofNL6DQ',
```
Comment (alexchiriac) [inline examples/oAuth2.0/balance-query-v2.js:16]: For this example, can we avoid using this exact token, and instead just put OverledgerSDK('') with a comment indicating what token needs to be placed in and where to get it from?

**Review by alexchiriac — COMMENTED** [review]
Left another comment

File: packages/overledger-core/src/OverledgerSDK.ts (hunk 3 above)
Comment (najla-quant) [reply packages/overledger-core/src/OverledgerSDK.ts:46]: Yes, I agree that testing the length wouldn't be the best way to choose between v1 or v2. Maybe adding a type argument in the constructor as a first argument ?

Comment (mahsa-bytt) [reply packages/overledger-core/src/OverledgerSDK.ts:46]: @alexchiriac @najla-quant  all the usage of this constructor is like : 
```
new OverledgerSDK(mappId, bpiKey, {
      dlts: [{ dlt: DltNameOptions.HYPERLEDGER_FABRIC }],
      provider: { network: fabricNetworkConnection },
    });
```
there's no place that we have for example just mappId+bpiKey without the SDKOptions. 
I can add the type of the args in my checking. 
Is that ok? or the whole overloading is not fine?

File: examples/oAuth2.0/balance-query-v2.js (hunk 2 above)
Comment (mahsa-bytt) [reply examples/oAuth2.0/balance-query-v2.js:40]: fixed.

File: packages/overledger-core/src/OverledgerSDK.ts (hunk 3 above)
Comment (najla-quant) [reply packages/overledger-core/src/OverledgerSDK.ts:46]: Could you do for example:

 constructor(authenticationType: string, ...args: any[]) {

}
?

Comment (alexchiriac) [reply packages/overledger-core/src/OverledgerSDK.ts:46]: Ahh, I see! The DLTs are mandatory, so yeah, then you wouldn't ever have 2 args! Ok :D

File: packages/overledger-core/src/OverledgerSDK.ts (hunk 4 above)
Comment (mahsa-bytt) [reply packages/overledger-core/src/OverledgerSDK.ts:59]: fixed.

File: packages/overledger-core/src/OverledgerSDK.ts (hunk 5 above)
Comment (mahsa-bytt) [reply packages/overledger-core/src/OverledgerSDK.ts:72]: fixed.

File: examples/oAuth2.0/balance-query-v2.js (hunk 6 above)
Comment (mahsa-bytt) [reply examples/oAuth2.0/balance-query-v2.js:16]: fixed.

File: examples/oAuth2.0/refresh-access-token.js (hunk 1 above)
Comment (mahsa-bytt) [reply examples/oAuth2.0/refresh-access-token.js:10]: fixed.

**Review by kashif-quant — APPROVED** [review]

File: packages/overledger-core/src/OverledgerSDK.ts (hunk 3 above)
Comment (mahsa-bytt) [reply packages/overledger-core/src/OverledgerSDK.ts:48]: @najla-quant what is authenticationType: string??
if I want to use overloading I can't do that, but I can change it like this:
`if (args.length === 3 && typeof args[0] === 'string' && typeof args[1] === 'string') {......`
I can check types too.

Comment (najla-quant) [reply packages/overledger-core/src/OverledgerSDK.ts:48]: For now, keep it like you did. If it becomes more complicated to test against length (jn case of new types of connection), we will figure out how to do it

Comment (najla-quant) [reply packages/overledger-core/src/OverledgerSDK.ts:48]: By authenticationType I meant the v1 or v2

Comment (mahsa-bytt) [reply packages/overledger-core/src/OverledgerSDK.ts:48]: @najla-quant ok, so for now I'll keep it like what I did.

**Pull request merged**